config/tools/config_generator.cpp

```diff
@@ -0,0 +1,897 @@
+        } else {
+            return create_repeatable_key_pair();
+        }
+    }
+
+    [[nodiscard]] auto config_generator::parse_value(const std::string& value,
+                                                     const bool keep_quotes)
+        -> value_t {
+        if(value[0] != '\"' && value[value.size() - 1] != '\"') {
+            if(value.find('.') == std::string::npos) {
+                const auto as_int = static_cast<size_t>(std::stoull(value));
+                return as_int;
+            }
+            const auto as_dbl = std::stod(value);
+            return as_dbl;
+        }
+
+        if(keep_quotes == false) {
+            const auto unquoted = value.substr(1, value.size() - 2);
+            return unquoted;
+        }
+        return value;
+    }
+
+    [[nodiscard]] auto config_generator::get_param_from_template_file(
+        const std::string option,
+        std::map<std::string, std::string>& config_map)
+        -> std::variant<size_t, double, std::string> {
+        auto it = config_map.find(option);
+        if(it != config_map.end()) {
+            value_t parsed_val = parse_value(it->second, false);
+            if(std::holds_alternative<size_t>(parsed_val)) {
+                return std::get<size_t>(parsed_val);
+            } else if(std::holds_alternative<double>(parsed_val)) {
+                return std::get<double>(parsed_val);
+            } else if(std::holds_alternative<std::string>(parsed_val)) {
+                return std::get<std::string>(parsed_val);
+            } else {
+                std::string error_msg
+                    = "Warning: Unrecognized type for param, " + option + ".";
+                std::cout << error_msg << std::endl;
+                return error_msg;
+            }
+        } else {
+            std::string error_msg
+                = "Warning: Could not find param, " + option + ".";
+            std::cout << error_msg << std::endl;
+            return error_msg;
+        }
+    }
+
+    void config_generator::set_param_to_config_file(const std::string key,
+                                                    const std::string value) {
+        m_new_config << key << "=" << '"' << value << '"' << '\n';
+    }
+
+    void config_generator::set_param_to_config_file(const std::string key,
+                                                    const size_t value) {
+        m_new_config << key << "=" << value << '\n';
+    }
+
+    void config_generator::set_param_to_config_file(const std::string key,
+                                                    const double value) {
+        m_new_config << key << "=" << value << '\n';
+    }
+
+    void config_generator::set_log_level(const std::string key,
+                                         std::string& log_level) {
+        if(find_value(key, log_level) == false) {
+            find_value(tmpl_universal_override_log_level, log_level);
+        }
+        if(log_levels.find(log_level) == log_levels.end()) {
```

Comment:
This check to see if the passed-in `log_level` is valid happens after we pass it through to `find_value()`; Is there a reason we shouldn't check first?

Reply:
Good point. I reversed the order.

DONE